contracts/src/l1_management/contracts/NetworkConfig.sol

```diff
@@ -234,4 +299,66 @@ contract NetworkConfig is Initializable, OwnableUpgradeable {
             additionalContracts: additional
         });
     }
+
+   /**
+     * @dev Records a contract upgrade
+     * @param name The name of the contract
+     * @param version The new version
+     * @param implementation The new implementation address
+     */
+    function recordUpgrade(
+        string calldata name,
+        string calldata version,
+        address implementation
+    ) external onlyOwner {
+        ContractVersion storage current = contractVersions[name];
+        string memory oldVersion = current.version;
+
+        contractVersions[name] = ContractVersion({
+            name: name,
+            version: version,
+            implementation: implementation
+        });
+
+        emit ContractUpgraded(name, oldVersion, version, implementation);
+    }
+
+    /**
+     * @dev Records multiple contract upgrades in a batch
+     * @param names Array of contract names
+     * @param versions Array of new versions
+     * @param implementations Array of new implementation addresses
+     */
+    function recordBatchUpgrade(
```

Comment:
I was imagining something simpler - string hardfork name and just emitting when an upgrade has happened; callable only by owner. This is fine i guess, but it doesn't bring in any trust benefits as implementations are still passed as calldata so we can say we've upgraded without any verifiability. 

I guess if we were to go down this approach it would be good to call this function with addresses of proxies, which are queried for their implementation address and then thats emitted - if we wanted the l2 to have some checks around contract correctness.